src/spectators.h

```diff
@@ -35,46 +35,31 @@ class SpectatorVec
 	}
 
 	void addSpectators(const SpectatorVec& spectators) {
-		const size_t size = vec.size();
 		for (Creature* spectator : spectators.vec) {
-			bool duplicate = false;
-			for (size_t i = 0; i < size; ++i) {
-				if (vec[i] == spectator) {
-					duplicate = true;
-					break;
-				}
-			}
-
-			if (!duplicate) {
-				vec.emplace_back(spectator);
+			auto it = std::find(vec.begin(), vec.end(), spectator);
+			if (it != end()) {
+				continue;
 			}
+			vec.emplace_back(spectator);
 		}
 	}
 
 	void erase(Creature* spectator) {
-		for (size_t i = 0, len = vec.size(); i < len; i++) {
-			if (vec[i] == spectator) {
-				Creature* tmp = vec[len - 1];
-				vec[len - 1] = vec[i];
-				vec[i] = tmp;
-				vec.pop_back();
-				break;
-			}
+		auto it = std::find(vec.begin(), vec.end(), spectator);
+		if (it == end()) {
+			return;
 		}
+		std::iter_swap(it, end());
```

Comment:
change
`std::iter_swap(it, end());`
for
`std::iter_swap(it, end() - 1);`

This is the problem that I had and I had not noticed it.
remember that the function exchanges two iterators in valid ranges, `end()` is not a valid range, therefore the vector does not change at all, instead if you do `end() - 1` it will return the last valid position

;)